### DISCUSSION TITLE
Is there a way to know the current type being using from a part display driver?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#9021] [Q&A] [answered]
I created an attachable content part. I created a display driver for the contact part so I can control how to render the part. 

I am trying to make the part render different data based on the content type its attached to. Like if the contact type its attached to is "A" then we render "A-DataSource". If the contact Type is "B" then render "B-DataSource".

Is there a way to determine the parent content type from the part model or the context? If not, is there a way to set displayType/Mode in the migration and then get the displayType/Mode in the driver?

I tried the following migration

```
_contentDefinitionManager.AlterTypeDefinition("NewListingsWidget", type => type.WithPart("ListingWidgetPart", part => part.WithDisplayMode("Latest")));
_contentDefinitionManager.AlterTypeDefinition("FeaturedListingsWidget", type => type.WithPart("ListingWidgetPart", part => part.WithDisplayMode("Featured")));
```

But when rending the `ListingWidgetPart` how can I access the display-mode?

**Comment by ns8482e** [comment, marked as answer]
Using display option you can define display mode in migration as per your code above.

You might be missing shapes - You need to define option shapes  -  `ListingWidgetPart_DisplayOption__Latest`  and `ListingWidgetPart_DisplayOption__Featured`  to have option to seen in admin, and detail display mode shapes `ListingWidgetPart_Display__Featured`, `ListingWidgetPart_Display__Latest` to render in UI

**Reply by MikeAlhayek** [reply]
@ns8482e thanks for your input. How can I get the display-mode from the driver directly?

The part `ListingWidgetPart` uses a custom driver to render the part. I want to fetch different datasets depending on the value of the display-mode before the views are called.

Here is my custom driver, I just can't find a way to get the display mode "Latest" or "Featured" from with in the driver.

```
public class ListingWidgetPartDisplayDriver : ContentPartDisplayDriver<ListingWidgetPart>
{
    public override async Task<IDisplayResult> DisplayAsync(ListingWidgetPart part, BuildPartDisplayContext context)
    {
        // figure out the display mode
        // fetch the corresponding data
        // create the shape from the fetched data
    }
}
```

**Reply by ns8482e** [reply]
I guess `ContentPartDisplayDriver` takes care of it for you, when you create an ShapeResult with `Initialize` in driver.

```cs
public class ListingWidgetPartDisplayDriver : ContentPartDisplayDriver<ListingWidgetPart>
{
    public override async Task<IDisplayResult> DisplayAsync(ListingWidgetPart part, BuildPartDisplayContext context)
    {
        return Initialize<ListingWidgetPartViewModel>("ListingWidgetPart", m =>
            {
                m.PropertyOne = part.PropertyOne;
            }).Location("Leading:before");
    }
}
```

If DisplayOption is set to `Featured` then above code should add an alternate that resolves to template `ListingWidgetPart-Featured.Display.cshtml` `

**Reply by MikeAlhayek** [reply]
@ns8482e the issue is not in the alternates. The issue is that I want to find the DisplayMode from the driver so I can control what data to pull to create the views. I am expanding more code here to clarify what I am trying to do.

```


public class ListingWidgetPartDisplayDriver : ContentPartDisplayDriver<ListingWidgetPart>
{
    public override async Task<IDisplayResult> DisplayAsync(ListingWidgetPart part, BuildPartDisplayContext context)
    {
    	// I need to get the Latest from the part or the context.
        List<RealEstatePropertySummaryPart> properties = await GetPropertiesAsync("Latest");

        var totalProperties = properties.Count();

        if (totalProperties == 0)
        {
            return null;
        }

        List<IDisplayResult> results = new List<IDisplayResult>();

        for (int i = 0; i < totalProperties; i++)
        {
        	var model = properties[i];
        	
            IDisplayResult property = Initialize<PropertySummaryViewModel>(nameof(PropertySummaryViewModel), viewModel =>
            {
                viewModel.Id = model.Id;
                viewModel.City = model.CityName;
                viewModel.State = model.StateName;
                //.....
            }).Location($"Property:{i}")
            .OnGroup(context.GroupId)

            results.Add(property);
        }

        return Combine(results);
    }


    public async Task<RealEstatePropertySummaryPart> GetPropertiesAsync(string displayMode)
    {
    	if(displayMode == "Featured")
    	{

    		return await PropertyProvider.GetFeaturedAsync();
    	}

    	return await PropertyProvider.GetLatestAsync();
    }
}
```

as you can see in the above code, the view that I am going to use is called `PropertySummaryViewModel` for any data. The only different is that I need to pull a different data set depending on the value of the DisplayMode. The method `GetPropertiesAsync` returns different data set depending on the provided displayMode.

How do I get the value of displayModel from within the driver?

**Reply by ns8482e** [reply]
From Driver, You can access display mode setting using `context.TypePartDefinition.GetSettings`  that will give you setting object

**Reply by ns8482e** [reply]
If you only care about display mode, then you can use extension method `context.TypePartDefinition.DisplayMode()`

**Reply by MikeAlhayek** [reply]
That indeed did it! Thank you so much!
